minutus-mruby-build-utils/src/mruby_dir.rs

```diff
@@ -0,0 +1,88 @@
+use std::{fs, io, path::Path};
+
+use fs_extra::dir::{copy as copy_dir, CopyOptions};
+
+/// Checks whether a directory is empty.
+///
+/// Consider the following scenario:
+///
+/// Suppose we need to build mruby. Although the mruby directory exists,
+/// it may contain no files at all.
+///
+/// In such cases, the build will inevitably fail.
+/// Knowing in advance whether the mruby directory is empty is crucial to
+/// ensuring the correctness of the build process.
+fn dir_is_not_empty<P: AsRef<Path>>(path: P) -> bool {
+    fs::read_dir(path)
+        .ok()
+        .and_then(|mut entries| entries.next())
+        .is_some()
+}
+
+fn try_to_copy_dir(from: &Path, to: &Path) -> io::Result<u64> {
+    let opts = CopyOptions::new().overwrite(true);
+
+    copy_dir(from, to, &opts).map_err(|e| {
+        let err_msg = format!(
+            "Failed to copy directory.
+      src dir: {from:?}, target dir: {to:?}
+      Err: {e}"
+        );
+        io::Error::other(err_msg)
+    })
+}
+
+/// Copies the "/path/to/mruby-src-dir" to "work_dir/mruby"
+///
+/// # Example
+///
+/// ```ignore
+/// let work_dir: PathBuf = env::var("OUT_DIR")?.into();
+/// let src_dir = Path::new("/tmp/mruby-3.4")
+///
+/// // src_dir => "{work_dir}/mruby"
+/// copy_to_mruby_dir(src_dir, &work_dir)?;
+/// ```
+pub(crate) fn copy_to_mruby_dir(src_dir: &Path, work_dir: &Path) -> io::Result<u64> {
+    let target_dir = work_dir.join("mruby");
+
+    if target_dir.exists() {
+        if dir_is_not_empty(&target_dir) {
+            println!("cargo:warning=Dir exists: {target_dir:?}");
+            return Ok(0);
+        }
+        // An existing but empty mruby directory is a clear sign of a failed or
+        // incomplete build.
+        // Perform cleanup at this stage to prevent issues with subsequent `fs::rename`
+        // operations.
+        fs::remove_dir_all(&target_dir)?
+    }
+    println!("cargo:warning=src dir: {src_dir:?}, target dir: {target_dir:?}");
```

Comment:
I think warn is for unusual cases, and info is for usual cases.

```suggestion
    if target_dir.exists() {
        if dir_is_not_empty(&target_dir) {
            println!("cargo:info=Dir exists: {target_dir:?}");
            return Ok(0);
        }
        // An existing but empty mruby directory is a clear sign of a failed or
        // incomplete build.
        // Perform cleanup at this stage to prevent issues with subsequent `fs::rename`
        // operations.
        fs::remove_dir_all(&target_dir)?
        println!("cargo:warn=removed existing target dir: {target_dir:?}");
    }
    println!("cargo:info=src dir: {src_dir:?}, target dir: {target_dir:?}");
```

Reply:
Unfortunately, `build.rs` currently does not support `cargo:info=` or `cargo:warn=`.  
Only version 1.77 and above support `cargo::error=`, but fortunately, older versions of Cargo do support `cargo:warning=`.  
For the code above, using `cargo:warning=` is appropriate—since the current stable version of Cargo does not support `cargo::debug=` or `cargo::info=`.

<https://github.com/rust-lang/cargo/issues/7037>
<https://doc.rust-lang.org/cargo/reference/build-scripts.html?highlight=warning#outputs-of-the-build-script>

Reply:
Although I don’t fully agree with your suggestion, I’ve still committed the related changes — just to make you happy and in hopes that this PR will be merged.

Reply:
Oh sorry, I thought cargo has `cargo:info=`. 

> I chose eprintln!("[INFO]")
> Since none of them are recognized by Cargo(without -vv)

I feel this is good 👍